pkg/frontend/mysql_cmd_executor.go

```diff
@@ -2158,6 +2179,94 @@ func (cwft *TxnComputationWrapper) Compile(requestCtx context.Context, u interfa
 	if err != nil {
 		return nil, err
 	}
+	// check if it is necessary to initialize the temporary engine
+	if cwft.compile.NeedInitTempEngine(cwft.ses.IfNeedInitTempEngine()) {
+		// 0. init memory-non-dist engine
+		ck := clock.DefaultClock()
+		if ck == nil {
+			ck = clock.NewHLCClock(func() int64 {
+				return time.Now().Unix()
+			}, math.MaxInt)
+		}
+		shard := logservicepb.DNShardInfo{
+			ShardID:   2,
+			ReplicaID: 2,
+		}
+		shards := []logservicepb.DNShardInfo{
+			shard,
+		}
+		dnAddr := "1"
+		dnStore := logservicepb.DNStore{
+			UUID:           uuid.NewString(),
+			ServiceAddress: dnAddr,
+			Shards:         shards,
+		}
+		ms, err := memorystorage.NewMemoryStorage(
+			mpool.MustNewZero(),
+			memorystorage.SnapshotIsolation,
+			ck,
+			memoryengine.RandomIDGenerator,
+		)
+		if err != nil {
+			return nil, err
+		}
+		mc := memorystorage.NewStorageTxnClient(
+			ck,
+			map[string]*memorystorage.Storage{
+				dnAddr: ms,
+			},
+		)
+		me := memoryengine.New(
+			requestCtx,
+			memoryengine.NewDefaultShardPolicy(
+				mpool.MustNewZero(),
+			),
+			func() (logservicepb.ClusterDetails, error) {
+				return logservicepb.ClusterDetails{
+					DNStores: []logservicepb.DNStore{
+						dnStore,
+					},
+				}, nil
+			},
+			memoryengine.RandomIDGenerator,
+		)
+		txnOp, err := mc.New()
+		if err != nil {
+			return nil, err
+		}
+
+		// 1. bind the temporary engine to the session
+		cwft.ses.SetTempEngine(me)
+		e := cwft.ses.GetStorage().(*engine.EntireEngine)
+
+		// 2. init temp-db to store temporary relations
+		err = e.TempEngine.Create(requestCtx, engine.TEMPORARY_DBNAME, txnOp)
+		if err != nil {
+			return nil, err
+		}
+
+		// 3. assign for compile, only for the first time needed
+		cwft.compile.SetTempEngine(e.TempEngine)
+		// 4. add auto_IncrementTable fortemp-db
+		txnop := cwft.proc.TxnOperator.(*client.EntireTxnOperator)
+
+		txnop.SetTemp(txnOp)
+		colexec.CreateAutoIncrTable(e, requestCtx, cwft.proc, engine.TEMPORARY_DBNAME)
+		txnOp.Commit(requestCtx)
+
+		txnOp2, err := mc.New()
+		if err != nil {
+			return nil, err
+		}
+		txnop.SetTemp(txnOp2)
+
+		txn := cwft.ses.txnCompileCtx.txnHandler
+		txn.SetTempEngine(e.TempEngine)
```

Comment:
How many places you need to put the temp engine into?

Reply:
Here is the necessary operation to make the temporary table support transactions, maybe the code is not written well enough.